pkg/sshd/server.go

```diff
@@ -28,58 +30,99 @@ const (
 	// ServerConfig is the context key for ServerConfig object.
 	ServerConfig string = "ssh.ServerConfig"
 
-	multiplePush string = "Another git push is ongoing"
+	multiplePush       string = "Another git push is ongoing"
+	builderKeyLocation        = "/var/run/secrets/api/auth/builder-key"
 )
 
+var errBuildAppPerm = errors.New("user has no permission to build the app")
+var errDirPerm = errors.New("Cannot change directory in file name.")
+var errDirCreatePerm = errors.New("Empty repo name.")
+
+// AuthKey authenticates based on a public key.
+func AuthKey(key ssh.PublicKey) (*ssh.Permissions, error) {
+	log.Info("Starting ssh authentication")
+	userInfo, err := controller.UserInfoFromKey(key)
+	if err != nil {
+		return nil, err
+	}
+
+	userInfo.Key = string(ssh.MarshalAuthorizedKey(key))
+	apps := strings.Join(userInfo.Apps, ", ")
+	log.Debug("Key accepted for user %s.", userInfo.Username)
+	perm := &ssh.Permissions{
+		Extensions: map[string]string{
+			"user":        userInfo.Username,
+			"fingerprint": userInfo.Fingerprint,
+			"apps":        apps,
+		},
+	}
+	return perm, nil
+}
+
+// Configure creates a new SSH configuration object.
+//
+// Config sets a PublicKeyCallback handler that forwards public key auth
+// requests to the route named "pubkeyAuth".
+//
+// This assumes certain details about our environment, like the location of the
+// host keys. It also provides only key-based authentication.
+// ConfigureServerSshConfig
+//
+// Returns:
+//  An *ssh.ServerConfig
+func Configure() (*ssh.ServerConfig, error) {
+	cfg := &ssh.ServerConfig{
+		PublicKeyCallback: func(m ssh.ConnMetadata, k ssh.PublicKey) (*ssh.Permissions, error) {
+			return AuthKey(k)
+		},
+	}
+	hostKeyTypes := []string{"rsa", "dsa", "ecdsa"}
+	pathTpl := "/var/run/secrets/deis/builder/ssh/ssh-host-%s-key"
+	for _, t := range hostKeyTypes {
+		path := fmt.Sprintf(pathTpl, t)
+
+		key, err := ioutil.ReadFile(path)
+		if err != nil {
+			log.Debug("Failed to read key %s (skipping): %s", path, err)
+			return nil, err
+		}
+		hk, err := ssh.ParsePrivateKey(key)
+		if err != nil {
+			log.Debug("Failed to parse host key %s (skipping): %s", path, err)
+			return nil, err
+		}
+		log.Debug("Parsed host key %s.", path)
+		cfg.AddHostKey(hk)
+	}
+	return cfg, nil
+}
+
 // Serve starts a native SSH server.
 //
 // The general design of the server is that it acts as a main server for
 // a Cookoo app. It assumes that certain things have been configured for it,
```

Comment:
also, you can remove the sentence following the mention of `Cookoo`, because this func now requires a `*ssh.ServerConfig` and the program won't build without one :)